### PR TITLE
[SPARK-44763][SQL] Fix a bug of promoting string as double in binary arithmetic with interval

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

The following query works on branch-3.5 or below, but fails on the latest master:

```
select concat(DATE'2020-12-31', ' ', date_format('09:03:08', 'HH:mm:ss')) + (INTERVAL '03' HOUR)
```
 
The direct reason is now we mark `cast(date as string)` as resolved during type coercion after changes https://github.com/apache/spark/pull/42089. As a result, there are two transforms from CombinedTypeCoercionRule
```
Rule ConcatCoercion Transformed
concat(2020-12-31,  , date_format(cast(09:03:08 as timestamp), HH:mm:ss, Some(America/Los_Angeles)))
to
concat(cast(2020-12-31 as string),  , date_format(cast(09:03:08 as timestamp), HH:mm:ss, Some(America/Los_Angeles)))

Rule PromoteStrings Transformed
(concat(cast(2020-12-31 as string),  , date_format(cast(09:03:08 as timestamp), HH:mm:ss, Some(America/Los_Angeles))) + INTERVAL '03' HOUR)
to
(cast(concat(cast(2020-12-31 as string),  , date_format(cast(09:03:08 as timestamp), HH:mm:ss, Some(America/Los_Angeles))) as double) + INTERVAL '03' HOUR)
```  
In previous releases, the second transform doesn't happen since `cast(2020-12-31 as string)`  was considered unresolved after the first transform. With rules `ResolveTimeZone` and `ResolveBinaryArithmetic`, the `Add` becomes `TimeAdd` before it is transformed by `CombinedTypeCoercionRule` again.
 
The fix is simple, the analyzer should not promote string as double in binary arithmetic with ANSI interval. The changes in [https://github.com/apache/spark/pull/42089](https://github.com/apache/spark/pull/42089.) are valid and we should keep it.

### Why are the changes needed?

Bug fix

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New UTs

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @xuanyuanking note that this is not a regression on branch-3.5, but I think it is nice to have in the coming release. 
You can make the call of whether to put it in branch 3.5 or not.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -1096,15 +1096,20 @@ object TypeCoercion extends TypeCoercionBase {
       }
     }
 
+    private def isIntervalType(dt: DataType): Boolean = dt match {
+      case _: CalendarIntervalType | _: AnsiIntervalType => true
```
Comment (amaliujia) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:1100]: Question: there is also `DayTimeIntervalType` and `YearMonthIntervalType` et.c and why they are not included in here?

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:1100]: My bad. I was looking at the Object. The class of `DayTimeIntervalType` and `YearMonthIntervalType` has already extended `AnsiIntervalType`

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:1100]: > Question: there is also DayTimeIntervalType and YearMonthIntervalType et.c and why they are not included in here?

They are of AnsiIntervalType

**Review by cloud-fan — APPROVED** [review]
good catch!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master.

**Pull request closed without merging**